Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs

```diff
@@ -467,44 +466,63 @@ public class ZitiRouterEvent {
 	}
 
 	public class ZitiServiceEvent {
-		internal IntPtr nativeRemovedList;
-		internal IntPtr nativeChangedList;
-		internal IntPtr nativeAddedList;
-		internal IntPtr ziti_ctx;
+
+		public ZitiServiceEvent(IntPtr zitiCtx) {
+			this.ziti_ctx = zitiCtx;
+		}
+		internal IntPtr nativeRemovedList {
+			set => removedList = createZitiServiceList(value);	
+        }
+		internal IntPtr nativeChangedList {
+			set => changedList = createZitiServiceList(value);
+        }
+		internal IntPtr nativeAddedList {
+			set => addedList = createZitiServiceList(value);
+		}
+		internal IntPtr ziti_ctx { get; }
 		internal ZitiIdentity id;
+		internal List<ZitiService> removedList;
+		internal List<ZitiService> changedList;
+		internal List<ZitiService> addedList;
 
 		public object Context { get; internal set; }
 
 		private IEnumerable<IntPtr> array_iterator(IntPtr arr) {
 			int index = 0;
 			while (true) {
-				IntPtr removedService = Native.API.ziti_service_array_get(arr, index);
+				IntPtr zitiService = Native.API.ziti_service_array_get(arr, index);
 				index++;
-				if (removedService == IntPtr.Zero) {
+				if (zitiService == IntPtr.Zero) {
 					break;
 				}
 
-				yield return removedService;
+				yield return zitiService;
 			}
 		}
 
-		public IEnumerable<ZitiService> Removed() {
-			foreach (IntPtr p in array_iterator(nativeRemovedList)) {
+		private List<ZitiService> createZitiServiceList(IntPtr p) {
+			List<ZitiService> servicesList = new List<ZitiService>();
+			foreach (IntPtr p in array_iterator(p)) {
 				ZitiService svc = new ZitiService(id, new ZitiContext(ziti_ctx), p);
+				servicesList.Add(svc);
+			}
+			return servicesList;
+		}
+
+		public IEnumerable<ZitiService> Removed() {
+			foreach (ZitiService svc in removedList) {
 				yield return svc;
 			}
 		}
 
 		public IEnumerable<ZitiService> Changed() {
-			foreach (IntPtr p in array_iterator(nativeChangedList)) {
-				ZitiService svc = new ZitiService(id, new ZitiContext(ziti_ctx), p);
+			foreach (ZitiService svc in changedList) {

```

Comment:
If the ZitiService objects are created inside the Added, Removed and Changed functions, it creates ZitiSevice objects for each native service, multiple times based on the number of times the enumerable object is called. So moved the code to setter of native service variable